llvm/lib/Target/AArch64/AArch64FrameLowering.cpp

```diff
@@ -2784,6 +2784,25 @@ struct RegPairInfo {
 
 } // end anonymous namespace
 
+void verify(SmallVectorImpl<RegPairInfo> &RegPairs) {
+  auto IsPPR = [](const RegPairInfo &c) { return c.Reg1 == RegPairInfo::PPR; };
+  auto PPRBegin = std::find_if(RegPairs.begin(), RegPairs.end(), IsPPR);
+  auto IsZPR = [](const RegPairInfo &c) { return c.Type == RegPairInfo::ZPR; };
+  auto ZPRBegin = std::find_if(RegPairs.begin(), RegPairs.end(), IsZPR);
+  assert(!(PPRBegin < ZPRBegin) &&
+         "Expected callee save predicate to be handled first");
+}
+
```

Comment:
Does this work when there are no PPR registers to save (PPRBegin == RegPairs.end())?

Additionally:
* You only need to check this when using a predicate reg to do the paired spill/fill instructions.
* nit: `verify` is quite a generic name, but the checks are rather specific. Maybe just inline this function where it is used?

Reply:
> Does this work when there are no PPR registers to save (PPRBegin == RegPairs.end())?

If there is no PReg to save then it should be fine to clobber PReg.
AFAIU the check is  to avoid to spill/store the PRegs after ZRegs and fill/load PRegs before ZRegs. If there is not PReg to spill and fill then the compiler should be fine.